### ISSUE TITLE
Drush updatedb / updb keeps running the same operations.

### ISSUE THREAD

**Issue by thomasdik1989** [drush-ops/drush#2762]
The sandbox variable is unavailable in the new thread which is created once 50% memory usage has been reached.

In case of large batches which are set based on an empty sandbox the batch will loop forever.

**Comment by claudiu-cristea** [comment]
I conform the bug.

**Comment by attrib** [comment]
Can't reproduce this with #3952 applied.

My test update code:
```
function as_base_update_8015(&$sandbox) {
  $entity_type_manager = \Drupal::entityTypeManager();
  $storage = $entity_type_manager->getStorage('node');

  if (!isset($sandbox['progress'])) {
    $query_count = $storage->getQuery();
    $sandbox['progress'] = 0;
    $sandbox['max'] = $query_count->count()->execute();
  }

  $ids = $storage->getQuery()
    ->sort('id')
    ->range($sandbox['progress'], 50)
    ->execute();

  $result = $storage->loadMultiple($ids);
  $sandbox['progress'] += count($result);

  $sandbox['#finished'] = empty($sandbox['max']) ? 1 : ($sandbox['progress'] / $sandbox['max']);
  \Drupal::logger('updb')->warning(print_r($sandbox, TRUE) . ' - ' . memory_get_usage());
}
```

Set my `memory_limit = 100M` to just force this and I have 1192 test nodes to load, which will exceed the memory limit of 100M just because of the static cache.

Output of `drush updb`
```
 --------- ----------- --------------- -------------
  Module    Update ID   Type            Description
 --------- ----------- --------------- -------------
  as_base   8014        hook_update_n   8014 -
 --------- ----------- --------------- -------------


 Do you wish to run the specified pending updates? (yes/no) [yes]:
 > yes

>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 50
>     [max] => 1192
>     [#finished] => 0.041946308724832
> )
>  - 49163784
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 100
>     [max] => 1192
>     [#finished] => 0.083892617449664
> )
>  - 50908408
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 150
>     [max] => 1192
>     [#finished] => 0.1258389261745
> )
>  - 52562096
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 200
>     [max] => 1192
>     [#finished] => 0.16778523489933
> )
>  - 54205976
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 250
>     [max] => 1192
>     [#finished] => 0.20973154362416
> )
>  - 55835328
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 300
>     [max] => 1192
>     [#finished] => 0.25167785234899
> )
>  - 57525832
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 350
>     [max] => 1192
>     [#finished] => 0.29362416107383
> )
>  - 59211376
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 400
>     [max] => 1192
>     [#finished] => 0.33557046979866
> )
>  - 60898912
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 450
>     [max] => 1192
>     [#finished] => 0.37751677852349
> )
>  - 62865112
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 500
>     [max] => 1192
>     [#finished] => 0.41946308724832
> )
>  - 64685760
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 550
>     [max] => 1192
>     [#finished] => 0.46140939597315
> )
>  - 66765640
>  [notice] Batch process has consumed in excess of 50% of available memory. Starting new thread
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 600
>     [max] => 1192
>     [#finished] => 0.50335570469799
> )
>  - 49625472
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 650
>     [max] => 1192
>     [#finished] => 0.54530201342282
> )
>  - 51606512
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 700
>     [max] => 1192
>     [#finished] => 0.58724832214765
> )
>  - 53699456
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 750
>     [max] => 1192
>     [#finished] => 0.62919463087248
> )
>  - 55363936
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 800
>     [max] => 1192
>     [#finished] => 0.67114093959732
> )
>  - 57076216
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 850
>     [max] => 1192
>     [#finished] => 0.71308724832215
> )
>  - 58900272
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 900
>     [max] => 1192
>     [#finished] => 0.75503355704698
> )
>  - 60869264
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 950
>     [max] => 1192
>     [#finished] => 0.79697986577181
> )
>  - 62516856
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 1000
>     [max] => 1192
>     [#finished] => 0.83892617449664
> )
>  - 64241392
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 1050
>     [max] => 1192
>     [#finished] => 0.88087248322148
> )
>  - 65909016
>  [notice] Batch process has consumed in excess of 50% of available memory. Starting new thread
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 1100
>     [max] => 1192
>     [#finished] => 0.92281879194631
> )
>  - 49130792
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 1150
>     [max] => 1192
>     [#finished] => 0.96476510067114
> )
>  - 50924400
>  [notice] Update started: as_base_update_8014
>  [warning] Array
> (
>     [progress] => 1192
>     [max] => 1192
>     [#finished] => 1
> )
>  - 52334728
>  [notice] Update completed: as_base_update_8014
 [success] Finished performing updates.```

As you can see there were two restarts (`>  [notice] Batch process has consumed in excess of 50% of available memory. Starting new thread`) and the sandbox content stayed the same.

**Comment by pfrenssen** [comment]
It looks like this can be closed, #3952 has been merged some time ago.
